File: src/main/java/com/demonwav/mcdev/platform/mixin/inspections/MixinClassReferenceInspection.java
package com.demonwav.mcdev.platform.mixin.inspections;

import com.intellij.psi.PsiAnnotation;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.PsiModifierList;
import com.intellij.psi.PsiType;
import com.intellij.psi.PsiTypeElement;
import com.intellij.psi.impl.source.PsiClassReferenceType;
import com.siyeh.ig.BaseInspection;
import com.siyeh.ig.BaseInspectionVisitor;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class MixinClassReferenceInspection extends BaseInspection {

    @Nls
    @NotNull
    @Override
    public String getDisplayName() {
        return "Reference to Mixin class!";
    }

    @NotNull
    @Override
    protected String buildErrorString(Object... infos) {
        PsiClass mixinClass = (PsiClass) infos[0];
        return mixinClass.getName() + " is a Mixin class, and cannot be referenced directly!";
    }

    @Nullable
    @Override
    public String getStaticDescription() {
        return "A Mixin class doesn't exist at runtime, and thus cannot be referenced directly.";
    }

    @Override
    public BaseInspectionVisitor buildVisitor() {
        return new BaseInspectionVisitor() {

            @Override
            public void visitTypeElement(PsiTypeElement typeElement) {
                final PsiType type = typeElement.getType();
                if (!(type instanceof PsiClassReferenceType)) {
                    return;
                }

                final PsiClass clazz = ((PsiClassReferenceType) type).resolve();
                if (clazz == null) {
                    return;
                }

                final PsiModifierList modifierList = clazz.getModifierList();
                if (modifierList == null) {
                    return;
                }
                final PsiAnnotation mixin = modifierList.findAnnotation("org.spongepowered.asm.mixin.Mixin");
                if (mixin == null) {
                    return;
                }

                registerError(typeElement, clazz);
            }
        };
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/mixin/inspection/MixinClassReferenceInspection.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.mixin.inspection

import com.demonwav.mcdev.platform.mixin.util.isAccessorMixin
import com.demonwav.mcdev.platform.mixin.util.isMixin
import com.demonwav.mcdev.util.findContainingClass
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.psi.JavaElementVisitor
import com.intellij.psi.PsiClassType
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.PsiTypeElement

class MixinClassReferenceInspection : MixinInspection() {

    override fun getStaticDescription() =
        "A Mixin class doesn't exist at runtime, and thus cannot be referenced directly."

    override fun buildVisitor(holder: ProblemsHolder): PsiElementVisitor = Visitor(holder)

    private class Visitor(private val holder: ProblemsHolder) : JavaElementVisitor() {

        override fun visitTypeElement(type: PsiTypeElement) {
            val classType = type.type as? PsiClassType ?: return
            val referencedClass = classType.resolve() ?: return

            if (!referencedClass.isMixin) {
                return
            }

            // Check if the the reference is a super Mixin
            val psiClass = type.findContainingClass() ?: return
            if (psiClass.isEquivalentTo(referencedClass) || psiClass.isInheritor(referencedClass, true)) {
                // Mixin class is part of the hierarchy
                return
            }

            // Check if the reference is an accessor Mixin
            if (referencedClass.isAccessorMixin) {
                return
            }

            holder.registerProblem(type, "Mixin class cannot be referenced directly")
        }
    }
}
